### ISSUE TITLE
should the train_test_split be placed outside the objective def (or have a defined random state)

### ISSUE THREAD

**Issue by ThomasBury** [optuna/optuna#2184]
Hi,

should the `train_test_split` be done outside the definition of the objective or at least the random_state fixed? If the objective is evaluated several times, the metric will then not computed using the same validation set and then the optimization is not comparing the same quantity over the different iterations?

```python
def objective(trial):
    data, target = sklearn.datasets.load_breast_cancer(return_X_y=True)
    train_x, valid_x, train_y, valid_y = train_test_split(data, target, test_size=0.25)
    dtrain = lgb.Dataset(train_x, label=train_y)

    param = {
...
    }

    gbm = lgb.train(param, dtrain)
    preds = gbm.predict(valid_x)
    pred_labels = np.rint(preds)
    accuracy = sklearn.metrics.accuracy_score(valid_y, pred_labels)
    return return accuracy
```

**Comment by HideakiImamura** [comment]
Thanks for the question. In my opinion, `train_test_split` should be called inside the objective function. The reasons are as follows.
First, Optuna's algorithm (Bayesian optimization) assumes that the value of the objective function, `y`, is stochastically determined for the value of the hyperparameter, `x`. In other words, the realization of `y` that we observe is the value determined by `x`, `f(x)`, plus noise. `y = f(x) + noise`. There are many factors that make up this noise, and one of the most important factors in machine learning hyperparameter optimization is the variability of the data set. A good hyperparameter is one that averages well over a wide range of dataset variability. (This is open to debate. There may be situations where you only want one that performs well on a particular data set.) The train-test split is also included in the dataset variability. We don't want a hyperparameter that performs well only on a particular train-test split. We want a hyperparameter that performs well on average on various train-test splits. Therefore, it would be better to do `train_test_split` every time in the objective function to get such hyperparameters.

**Comment by ThomasBury** [comment]
Hi,

Thank you for your reply, I mostly agree.  However, considering a different train and validation split for each trial (each combination of HP), then one is not comparing the error on the same set. This means that one is evaluating both the knowledge uncertainty and data uncertainty which can lead to stochastic variability of the score(s).  Assuming that there is an overall best hyper-parameters combination, minimizing the knowledge uncertainty (meaning that the model could not perform better than with those HP), the score on a specific split might be worst than on another one only by "bad luck" (meaning because the data uncertainty is larger on this particular split than on the others).

Cross-validation, on the other hand, allows comparing/scoring every trial (combination) of hyper-parameters in a consistent fashion, fixing a common gauge for all the HP trials and allows data uncertainty smoothing at the same time. Using the same train and validation sets so that the score values are on the same scale and repeat for each folds to average out the intrinsic noise (aka data uncertainty). All the HP combinations are then scored in a fair way, the variability coming mostly from the knowledge(model) uncertainty. [A better explanation can be found here](https://arxiv.org/pdf/2006.10562.pdf)

For this reason, I'd fix the seed of train_test_split and use cross-validation if one wants to handle data uncertainty.

Btw, Optuna is for sure one of the best HP optimization packages, a great gift to the open-source community.
KR
